src/devices/Board/I2cBusExtensions.cs

```diff
@@ -44,5 +44,112 @@ public static List<int> PerformBusScan(this I2cBus bus, int lowest = 0x3, int hi
 
             return ret;
         }
+
+        /// <summary>
+        /// Performs a scan on the I2C bus, returning the addresses for all connected devices.
+        /// </summary>
+        /// <param name="bus">The bus to scan</param>
+        /// <param name="progress">Progress feedback provider. Receives scan progress in percent</param>
+        /// <param name="lowestAddress">The lowest address to scan. Default 0x03</param>
+        /// <param name="highestAddress">The highest address to scan. Default 0x77</param>
+        /// <returns>A list of bus addresses that are in use, an empty list if no device was found</returns>
+        /// <remarks>This method should never throw an exception. Bus scanning may interfere with normal device operation,
+        /// so this should not be done while devices are being used.</remarks>
+        public static (List<int> FoundDevices, int LowestAddress, int HighestAddress) PerformBusScan(this I2cBus bus, IProgress<float>? progress, int lowestAddress = 0x3, int highestAddress = 0x77)
+        {
+            List<int> addresses = new();
+            int numberOfDevicesToScan = highestAddress - lowestAddress;
+            float currentPercentage = 0;
+            float stepPerDevice = 100.0f / numberOfDevicesToScan;
+            if (lowestAddress < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(lowestAddress));
+            }
+
+            if (highestAddress < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(highestAddress));
+            }
+
+            if (numberOfDevicesToScan <= 0)
+            {
+                return (addresses, 0, 0);
+            }
+
+            for (int addr = lowestAddress; addr <= highestAddress; addr++)
+            {
+                try
+                {
+                    using (var device = bus.CreateDevice(addr))
```

Comment:
NIT: I know I haven't commented on this in the past, but we usually try to avoid using `var` keyword unless the type can be inferred by the assignment like for example: `var x = new Device();` so please try to switch all of these other uses where you are not calling the constructor directly and instead use the concrete type. There is a roslyn analyzer that flags this and it does provide a fix-all solution, so perhaps we should go and run this code-fixer all over our source code in a subsequent PR.